rust-runtime/aws-smithy-async/src/future/pagination_stream.rs

```diff
@@ -3,16 +3,51 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-//! Utility to drive a stream with an async function and a channel.
+//! Types to support stream-like operations for paginators.
 
 use crate::future::rendezvous;
-use futures_util::StreamExt;
 use pin_project_lite::pin_project;
+use std::fmt;
+use std::future::poll_fn;
 use std::future::Future;
 use std::pin::Pin;
 use std::task::{Context, Poll};
-use tokio_stream::{Iter, Once, Stream};
 
+pub mod collect;
+
+/// A wrapper around [`FnStream`].
+///
+/// This type provides the same set of methods as [`FnStream`], but that is meant to be used
+/// internally and not by external users.
+#[derive(Debug)]
+pub struct PaginationStream<Item>(FnStream<Item>);
+
+impl<Item> PaginationStream<Item> {
+    /// Creates a `PaginationStream` from the given [`FnStream`].
+    pub fn new(stream: FnStream<Item>) -> Self {
+        Self(stream)
+    }
+
+    /// Consumes and returns the next `Item` from this stream.
+    pub async fn next(&mut self) -> Option<Item>
+    where
+        Self: Unpin,
+    {
+        self.0.next().await
+    }
+
+    /// Consumes this stream and gathers elements into a collection.
+    pub async fn collect<T: collect::sealed::Collectable<Item>>(self) -> T {
```

Comment:
what do you think about `try_collect` to do the `Result` flattening? I think that's more intuitive to use since it doesn't need a crazy turbofish. It will also remove the need for the Collectible trait

Reply:
(collect would return `Vec<Result`, try collect returns `Result<Vec`)

Reply:
Or we could add this method in addition and keep the generic 🤷🏻 

I just don't like typing `.collect::<Result<Vec<_>, _>()`

Reply:
Thanks for the suggestion! Addressed in  `43318e7`.

Reply:
Thanks for the suggestion! Addressed in  43318e7.